### PR TITLE
feat(grammar) rawstring

### PR DESCRIPTION
**Pull request by pluveto**
Now we can use \` to construct string so that we don't need to repeatedly write escape slashes.

_Fix KAG-370_

### CODE REVIEW

**Comment by chronolaw** [comment]
Could we add one more regex test case for this feature? such as `^/\d+/test$`.

**Comment by pluveto** [comment]
> Could we add one more regex test case for this feature? such as `^/\d+/test$`.

More test cases added

File: src/atc_grammar.pest [hunk 1]
```diff
@@ -17,6 +17,8 @@ str_inner = _{ (str_esc | str_char)* }
 str_char = { !("\"" | "\\") ~ ANY }
 str_esc = { "\\" ~ ("\"" | "\\" | "n" | "r" | "t") }
 
+rawstr_literal = ${"`" ~ rawstr_char* ~ "`"}
```
Comment (gszr) [inline src/atc_grammar.pest:20]:
```suggestion
rawstr_literal = ${ "`" ~ rawstr_char* ~ "`" }
```

File: src/atc_grammar.pest [hunk 2]
```diff
@@ -17,6 +17,8 @@ str_inner = _{ (str_esc | str_char)* }
...
+rawstr_char = { !"`" ~ ANY}
```
Comment (gszr) [inline src/atc_grammar.pest:21]:
```suggestion
rawstr_char = { !"`" ~ ANY }
```

**Review by gszr — COMMENTED** [review]
Very minor style correction - to align with other rules.

**Comment by dndx** [comment]
Great work, although I have some question regarding the usage of ` for raw strings. Have we considered any other alternatives?

Reason being I have rarely seen ` being used for raw strings in other languages, so am a bit hesitant.

**Comment by pluveto** [comment]
> Great work, although I have some question regarding the usage of ` for raw strings. Have we considered any other alternatives?
> 
> Reason being I have rarely seen ` being used for raw strings in other languages, so am a bit hesitant.

Any suggestion?

**Comment by gszr** [comment]
What about `'` @dndx?

**Comment by chronolaw** [comment]
> What about `'` @dndx?

Perhaps `'` is not a good choice. The user will use the `'` to quote expressions in the command line.

**Comment by gszr** [comment]
> > What about `'` @dndx?
> 
> Perhaps `'` is not a good choice. The user will use the `'` to quote expressions in the command line.

` also has the potential to cause issues in command line usage, as it is reserved for command substitution.

**Comment by chronolaw** [comment]
> how about simply use `_rstart_` `_rend_` as boundary?

In Lua we have `[[ ... ]]`, in C++ we have `R(...)`, we should find an acceptable form.

**Comment by pluveto** [comment]
quotes/parens are common in raw string which means we need to implement escape at least two chars in a raw string literal value.

**Comment by hbagdi** [comment]
@pluveto @gszr @chronolaw Are we still pursuing this one?

**Comment by chronolaw** [comment]
> @pluveto @gszr @chronolaw Are we still pursuing this one?

We still have some discussion on it.

**Comment by dndx** [comment]
@pluveto What about `r""`, just like Python. I think it should not conflict with our grammar rules.

**Comment by chronolaw** [comment]
> @pluveto What about `r""`, just like Python. I think it should not conflict with our grammar rules.

It seems that it is still hard to use `r""` in normal string, like "http.path=r\"...\""

**Comment by pluveto** [comment]
How about rust style:

```rs
let s = r#"'"""''"\/"# // '"""''"\/
```

**Review by dndx — APPROVED** [review]

File: src/parser.rs [hunk 3]
```diff
@@ -88,6 +88,18 @@ impl ATCParser {
 
         Ok(s)
     }
+    fn rawstr_literal(input: Node) -> ParseResult<String> {
```
Comment (chronolaw) [inline src/parser.rs:91]:
```suggestion

    fn rawstr_literal(input: Node) -> ParseResult<String> {
```

**Pull request merged**